Tools/cases_generator/generate_cases.py

```diff
@@ -105,20 +109,52 @@ class Formatter:
 
     stream: typing.TextIO
     prefix: str
-
-    def __init__(self, stream: typing.TextIO, indent: int) -> None:
+    emit_line_directives: bool = False
+    lineno: int  # Next line number, 1-based
+    filename: str  # Slightly improved stream.filename
+    nominal_lineno: int
+    nominal_filename: str
+
+    def __init__(
+            self, stream: typing.TextIO, indent: int, emit_line_directives: bool = False
+    ) -> None:
         self.stream = stream
         self.prefix = " " * indent
+        self.emit_line_directives = emit_line_directives
+        self.lineno = 1
+        # Make filename more user-friendly and less platform-specific
+        filename = self.stream.name.replace("\\", "/")
+        if filename.startswith("./"):
+            filename = filename[2:]
+        if filename.endswith(".new"):
+            filename = filename[:-4]
+        self.filename = filename
+        self.nominal_lineno = 1
+        self.nominal_filename = filename
 
     def write_raw(self, s: str) -> None:
         self.stream.write(s)
+        newlines = s.count("\n")
+        self.lineno += newlines
+        self.nominal_lineno += newlines
 
     def emit(self, arg: str) -> None:
         if arg:
             self.write_raw(f"{self.prefix}{arg}\n")
         else:
             self.write_raw("\n")
 
+    def set_lineno(self, lineno: int, filename: str) -> None:
+        if self.emit_line_directives:
+            if lineno != self.nominal_lineno or filename != self.nominal_filename:
+                self.emit(f'#line {lineno} "{filename}"')
+                self.nominal_lineno = lineno
+                self.nominal_filename = filename
+
+    def reset_lineno(self) -> None:
+        if self.lineno != self.nominal_lineno or self.filename != self.nominal_filename:
+            self.set_lineno(self.lineno + 1, self.filename)
```

Comment:
Nit suggestion,
How about printing begin of implementation and the end of implementation more visually recognize?

```suggestion
            self.emit(f'#line {self.nominal_lineno} {self.nominal_filename} -> end of implementaion')
            self.set_lineno(self.lineno + 1, self.filename)
```

Reply:
I think the text in `"..."` must be exactly the filename. See e.g. https://gcc.gnu.org/onlinedocs/cpp/Line-Control.html. I'm not even sure if a comment is allowed on the same line.

Also, there will be many such directives, in a generated file that almost nobody will ever look at. Why do you want to add the "begin/end of implementation" markers? It should be clear what's what from the filename itself.

Reply:
> I think the text in "..." must be exactly the filename. See e.g. https://gcc.gnu.org/onlinedocs/cpp/Line-Control.html. I'm not even sure if a comment is allowed on the same line.

Ah I see, it that case, we should follow the GCC rule. I didn't know that.

Reply:
> Why do you want to add the "begin/end of implementation" markers? It should be clear what's what from the filename itself.

Just for visual identification of code block. But not absolutely needed.

Reply:
Looks like MSVC also support the feature
https://learn.microsoft.com/en-us/cpp/preprocessor/hash-line-directive-c-cpp?view=msvc-140

Reply:
It’s a standard C and C++ feature, I just linked to the simplest docs I found.